readers/api/transport.go

```diff
@@ -28,7 +28,8 @@ const (
 
 var (
 	errInvalidRequest     = errors.New("received invalid request")
-	errUnauthorizedAccess = errors.New("missing or invalid credentials provided")
+	errUnauthorizedAccess = errors.New("missing credentials provided")
+	errAccessDenied       = errors.New("invalid credentials provided")
```

Comment:
This should be kept consistent all over the code. If there is a difference between missing and invalid credentials (AuthN/AuthZ), then it should be detected in other places, and consequently 401/403 (both of them) codes should be used.

Reply:
There are 3 cases: 

- Missing token — 401
- Expired token (or maybe invalid by another reason) — 401
- Valid token, but access to some resource with this token is denied — 403

In this example, the current error message ("invalid credentials provided") maybe is a little bit confusing and should be changed to something else (like "permission denied")

Also maybe I should rename `errAccessDenied` to `errPermissionDenied`

Reply:
> There are 3 cases:
>
>Missing token — 401
>Expired token (or maybe invalid by another reason) — 401
>Valid token, but access to some resource with this token is denied — 403

@sprql exactly - well understood. Name of the error does not bother me much - `errAccessDenied` is OK and it relates to `403` rather than `401`, i.e. it is self-documenting enough.

What I wanted to check with you is did you cover all the code code base and identified where one of these 3 cases might appear and well differentiated between them - because here I can see `403` practically appearing just at one place in the whole Mainflux, according to your changes.

I did not review whole PR in details - it's bit difficult to review, this is why I am asking (looks to me you know your way areound Mainflux now and you know what you are doing, which is very good).

Anyway - thanks for this PR, it is very nice and think that you are really on a good track here for a nice improvement. Just please do not forget to update Swagger files as well.

@nmarcetic and @dusanb94 should review this one as well.

Reply:
I agree with this approach. This will later translate to handling errors from AuthN and AuthZ services (at the moment, only AuthN exists), where errors from AuthN will cause `Forbidden`, and errors from AuthZ will be `Unauthorized`.

Reply:
Great, we have general alignment of maintainers then.

@sprql please double-check the PR, ad final touches prepare it for meging. Please do not forget to add Swagger changes.

After that PR will be merged.

Reply:
@drasko, currently, the code mainly uses the `errUnauthorizedAccess`, both for AuthN & AuthZ errors, therefore, we will need to enter a new error type in these places. 

One option is to use error `errUnauthorizedAccess` for AuthN, and enter `errPermissionDenied` for AuthZ. But 'errUnauthorizedAccess' is misleading because of its name. 

Alternatively, we can enter `errUnauthenticated' for AuthN errors (like in gRPC).

In other words, the following options are possible:
`errUnauthorizedAccess` (401), `errPermissionDenied` (403)
or `errUnauthenticated` (401), `errUnauthorizedAccess` (403)
or `errUnauthenticated` (401), `errPermissionDenied` (403)

What do you think about it?

Reply:
Good question... I would avoid option 2, as this can be misleading (`errUnauthorizedAccess` might be related to 401 when here should be related to 403).

Between options 1 and 3, probably 3 is the least ambigous, although we could probably understant 1 as well now that we know what this is about...

I would personally vote for 3, to remove any abiguity, but I am also interested what others think. @manuio @nmarcetic @dusanb94 @mteodor @blokovi

Reply:
I also vote for 3

Reply:
I updated PR using option 3

Reply:
@sprql please do ot hate me, but after looking at PR I realized that actually option 2 would be more apropriate for Mainflux.

Actually, in Mainflux we have AuthN and AuthZ services, and I think that we should align errors to these functionalities, and not to outdated and missleading HTTP status names.

So I propose option 2, with a slight change: instad of `errUnauthorizedAccess` for `403` just use name `errUnauthorized`, to align it to it's 401 counterpart (`errUnauthenticated`).

Sorry for this, but I think now that everything is almost done, it will mostly be find-replace job :).

Reply:
@drasko I've updated, could you look at this again, I also updated error messages